sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -2654,6 +2659,13 @@ public PCollection<KafkaRecord<K, V>> expand(PCollection<KafkaSourceDescriptor>
         if (getRedistributeNumKeys() == 0) {
           LOG.warn("This will create a key per record, which is sub-optimal for most use cases.");
         }
+        // is another check here needed for with commit offsets
+        if (isCommitOffsetEnabled() || configuredKafkaCommit()) {
```

Comment:
for configuredkafkacommit regardless of rediistribute the checkpoint will be ahead of processed records.

for isCommitOffsetEnabled the offset may be ahead or behind of the processing (since it's a separate pipeline branch) but it will be consistent on drain.

Reply:
Your explanation for isCommitOffsetEnabled makes sense,

For configuredKafkaCommit, wouldn't the checkpoint always be behind processed records? We read messages, and then auto commit (default is every 5 seconds), so in between those 5 seconds, processing moves ahead of the last committed checkpoint?

Reply:
kafka autocommit can be ahead of records processed because the autocommit occurs when the consumer reads the data but the read data may be dropped if the commit to dataflow backend fails.  Dataflow won't lose data in this case since it will reread from the previous offset which it stores internally but the auto commit offset will be further ahead than that offset.

That is true in all cases but if the pipeline is drained at that point the commitOffsetEnabled would be eventually correct matching what was actually processed by the pipeline while the configuredKafkaCommit would remain incorrect. 

Separately should this log only be if there is redistribute withDuplicates allowed?  If not, the redistribute is just like a reshuffle which will have normal semantics.

Reply:
That is true. I've updated to add that check.